src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,541 @@
+	MultiConnection *connection;
+
+	connection = StartNodeUserDatabaseConnection(flags, hostname, port, user, database);
+
+	FinishConnectionEstablishment(connection);
+
+	return connection;
+}
+
+
+/*
+ * StartNodeUserDatabaseConnection() initiates a connection to a remote node.
+ *
+ * If user or database are NULL, the current session's defaults are used. The
+ * following flags influence connection establishment behaviour:
+ * - NEW_CONNECTION - it is permitted to establish a new connection
+ * - CACHED_CONNECTION - it is permitted to re-use an established connection
+ * - SESSION_LIFESPAN - the connection should persist after transaction end
+ * - FOR_DML - only meaningful for placement associated connections
+ * - FOR_DDL - only meaningful for placement associated connections
+ * - CRITICAL_CONNECTION - transaction failures on this connection fail the entire
+ *   coordinated transaction
+ *
+ * The returned connection has only been initiated, not fully
+ * established. That's useful to allow parallel connection establishment. If
+ * that's not desired use the Get* variant.
+ */
+MultiConnection *
+StartNodeUserDatabaseConnection(uint32 flags, const char *hostname, int32 port, const
+								char *user, const char *database)
+{
+	ConnectionHashKey key;
+	ConnectionHashEntry *entry = NULL;
+	MultiConnection *connection;
+	MemoryContext oldContext;
+	bool found;
+
+	strlcpy(key.hostname, hostname, MAX_NODE_LENGTH);
+	key.port = port;
+	if (user)
+	{
+		strlcpy(key.user, user, NAMEDATALEN);
+	}
+	else
+	{
+		strlcpy(key.user, CurrentUserName(), NAMEDATALEN);
+	}
+	if (database)
+	{
+		strlcpy(key.database, database, NAMEDATALEN);
+	}
+	else
+	{
+		strlcpy(key.database, get_database_name(MyDatabaseId), NAMEDATALEN);
+	}
+
+	if (CurrentCoordinatedTransactionState == COORD_TRANS_NONE)
+	{
+		CurrentCoordinatedTransactionState = COORD_TRANS_IDLE;
+	}
+
+	/*
+	 * Lookup relevant hash entry. We always enter. If only a cached
+	 * connection is desired, and there's none, we'll simply leave the
+	 * connection list empty.
+	 */
+
+	entry = hash_search(ConnectionHash, &key, HASH_ENTER, &found);
+	if (!found)
+	{
+		entry->connections = NIL;
+	}
+
+	if (flags & CACHED_CONNECTION)
+	{
+		ListCell *connectionCell = NULL;
+
+		/* check connection cache for a connection that's not already in use */
+		foreach(connectionCell, entry->connections)
+		{
+			connection = (MultiConnection *) lfirst(connectionCell);
+
+			/* don't return claimed connections */
+			if (!connection->claimedExclusively)
+			{
+				if (flags & SESSION_LIFESPAN)
+				{
+					connection->sessionLifespan = true;
+				}
+				connection->activeInTransaction = true;
+
+				/*
+				 * Check whether we're right now allowed to open new
+				 * connections. A cached connection counts as new if it hasn't
+				 * been used in this transaction.
+				 *
+				 * FIXME: This should be removed soon, once all connections go
+				 * through this API.
+				 */
+				if (!connection->activeInTransaction &&
```

Comment:
Maybe omit the activeInTransaction flag, it doesn't seem to be used (always true here?) and not quite clear how we could know that here.